lib/plugin.lazy.js

```diff
@@ -0,0 +1,109 @@
+import VueLib from 'vue'
+
+<% if (options.lazy.injectMock) { %>
+const apiMethods = <%= JSON.stringify(options.lazy.mockApiMethods)%>
+let delayedCalls = []
+let SentryMock = {}
+<% } %>
+let sentryReadyResolve
+
+export default async function SentryPlugin(ctx, inject) {
+  <% if (options.lazy.injectMock) { %>
+  apiMethods.forEach(key => {
+    SentryMock[key] = (...args) => delayedCalls.push([key, args])
+  })
+
+  window.addEventListener('error', SentryMock.captureException)
+
+  // Inject Sentry mock to the context as $sentry
+  inject('sentry', SentryMock)
+  ctx.$sentry = SentryMock
+  <% } %>
+
+  const loadSentryHook = () => loadSentry(ctx, inject)
+
+  <% if (options.lazy.injectLoadHook) { %>
+  inject('sentryLoad', loadSentryHook)
+  ctx.$sentryLoad = loadSentryHook
+  <% } else { %>
+  window.onNuxtReady(loadSentryHook)
+  <% } %>
+
+  const sentryReadyPromise = new Promise(resolve => {
+      sentryReadyResolve = resolve
+    })
+
+  const sentryReady = () => sentryReadyPromise
+
+  inject('sentryReady', sentryReady)
+  ctx.$sentryReady = sentryReady
+}
+
+async function loadSentry(ctx, inject) {
+  <%
+  const magicComments = [`webpackChunkName: '${options.lazy.chunkName}'`]
+  if (options.lazy.webpackPrefetch) {
+    magicComments.push('webpackPrefetch: true')
+  }
+  if (options.lazy.webpackPreload) {
+    magicComments.push('webpackPreload: true')
+  }
+  %>
+  const Sentry = await import(/* <%= magicComments.join(', ') %> */ '@sentry/browser')
+  const { <%= Object.keys(options.integrations).join(', ') %> } = await import(/* <%= magicComments.join(', ') %> */ '@sentry/integrations')
+
+  <% if (options.initialize) { %>// Initialize sentry
+  const config = <%= serialize(options.config) %>
+  config.integrations = [
+    <%= Object.entries(options.integrations).map(([name, integration]) => {
+      if (name === 'Vue') {
+        return `new ${name}({Vue: VueLib, ...${serialize(integration)}})`
+      }
+
+      const integrationOptions = Object.entries(integration).map(([key, option]) => {
+        typeof option === 'function'
+        ? `${key}:${serializeFunction(option)}`
+        : `${key}:${serialize(option)}`
+      })
+
+      return `new ${name}({${integrationOptions.join(',')}})`
+    }).join(',\n      ')%>
+  ]
+
+  Sentry.init(config)
+  <% } %>
+
+  <% if (options.lazy.injectMock) { %>
+  window.removeEventListener('error', SentryMock.captureException)
+
+  delayedCalls.forEach(([methodName, args]) => Sentry[methodName].apply(Sentry, args))
+  <% } %>
+
+  // Inject Sentry to the context as $sentry
+  sentryReadyResolve(Sentry)
+  inject('sentry', Sentry)
+  inject('sentryReady', () => Promise.resolve(Sentry))
```

Comment:
It's not possible to re-inject something that was already injected - see https://github.com/nuxt/nuxt.js/blob/6ab65fbf4f2b0d2348363c43f5c2fe81ef47cbc6/packages/vue-app/template/index.js#L185-L188

This will do nothing which is a problem because the potential code will keep using the mock (and crash) even after sentry has loaded.

Reply:
Ah damn, I should've know that. Its easily solvable except for the mock part. Either we should make sentry a function too or just skip the mock and tell people to always call+await `sentryReady`.

Eg we could reuse the mocked object reference, but then we have to assign the sentry moduel with `Object.assign(SentryMock, Sentry)`. That feels a bit hacky, but on the other side if it works it works. Wdyt?

Reply:
I found another way, we can just temporarily inject the SentryMock ourselves without calling Nuxt's inject. Then if the real Sentry is ready we use the inject call

Reply:
Those approaches sound hacky indeed.
Not sure what to do... A Proxy object would be the best but it's not supported by IE11.

Reply:
Yeah, Proxies are a no go imo until Vue3. I think this solution will work without having to change `$sentry` to a function